P5/Source/Specs/att.cmc.xml

```diff
@@ -0,0 +1,142 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright TEI Consortium.
+Dual-licensed under CC-by and BSD2 licences
+See the file COPYING.txt for details.
+-->
+<?xml-model href="http://jenkins.tei-c.org/job/TEIP5-dev/lastSuccessfulBuild/artifact/P5/release/xml/tei/odd/p5.nvdl" type="application/xml" schematypens="http://purl.oclc.org/dsdl/nvdl/ns/structure/1.0"?>
+<classSpec xmlns="http://www.tei-c.org/ns/1.0" xmlns:sch="http://purl.oclc.org/dsdl/schematron" ident="att.cmc" type="atts" module="tei">
+  <gloss xml:lang="en" versionDate="2023-11-24">computer mediated communication</gloss>
+  <desc xml:lang="en" versionDate="2021-09-07">provides attributes
+  categorizing how the element content was created in a CMC
+  environment.</desc>
+  <attList>
+    <attDef ident="generatedBy" mode="add" module="cmc" usage="opt">
+      <gloss xml:lang="en" versionDate="2021-09-07">generated by</gloss>
+      <desc xml:lang="en" versionDate="2021-09-07">categorizes
+      how the content of an element was generated in a CMC
+      environment.</desc>
+      <datatype minOccurs="1" maxOccurs="1">
+        <dataRef key="teidata.enumerated"/>
+      </datatype>
+      <constraintSpec scheme="schematron" ident="CMC_generatedBy_within_post">
+        <!-- WARNING: This constraint will NOT work using the skeleton
+             implementation (which we, TEI, currently use) due to a
+             bug in that system. Thus do not expect this constraint to
+             work until issue 1981 in TEIC TEI repo is fixed (e.g., by
+             PR 2509). —Syd, 2023-11-27 -->
+        <constraint>
+          <sch:rule context="@generatedBy">
```

Comment:
@sydb — you note above that this schematron does not work using the skeleton implementation (but it's not clear to me why this constraint won't work in skeleton?). 

Regardless, I think this would be a simple fix: 

```
<sch: rule context="tei:*[@generatedBy]">
<sch:assert test="ancestor-or-self::tei:post"><!--[...]--></sch:assert>
</sch:rule>
```

Reply:
@sydb We should change the rule context in order to match the prose of the Guidelines chapter and the spec, where we introduce `@generatedBy` as an attribute that appears on a `<post>` and/or its descendants. Considering the hosting element as an ancestor is unhealthy for processing.

Reply:
Yes, @joeytakeda, I have double-checked and you are correct, it is just the context being an attribute node that the Skeleton implementation fails to do properly. Thus the fix you have suggested should do exactly what we want until Skeleton is fixed (or maybe we use Mausatron internally). We should, of course, also add a test to detest. I expect to check in both of these changes tonight.
@ebeshero — Sorry, have not figured out what you are suggesting, here.

Reply:
Egads! **This is a release-blocker**

The test in question tries to enforce the rule that the `@generatedBy` attribute, although allowed on many different elements, is only allowed to occur *inside* a `<post>`. The new version of the test looks quite reasonable:

~~~xml
  <constraintSpec scheme="schematron" ident="CMC_generatedBy_within_post">
    <constraint>
      <sch:rule context="tei:*[@generatedBy]">
        <sch:assert test="ancestor-or-self::tei:post">
          The @generatedBy attribute is for use within a &lt;post> element.
        </sch:assert>
      </sch:rule>
    </constraint>
  </constraintSpec>
~~~

However, it causes our build process to fail. Why? Because there are examples (i.e., `<teix:egXML>` elements) that demonstrate the use of `@generatedBy` but do not have the ancestor `<post>`. This happens 3 times in 2 examples, both in the att.cmc.xml file, currently lines 114 & 125. (Embarrassingly, this is the same file in which the constraint above is defined.) You can find the two examples in p5.xml or p5subset.xml with the XPaths `//tei:exemplum[@type eq 'usenet']` and `//tei:exemplum[@type eq 'wikitalk']`.

Possible solutions that jump to my brain, in my (current) order of preference:
 * Re-write the examples so they include the surrounding `<post>` element.
 * Re-write the test so it is a special case that does not fire if it is inside a `<tei:exemplum>` that has `@type = ('wikitalk','usenet')`.
 * Re-write the test so it does not fire if it is inside a `<teix:egXML>`.
 * Drop the test entirely.
 * Re-write our build process so it does not test the contents of `<egXML>` by changing the namespaces from teix: to tei: and running the normal validation.

The new constraint and a test for it have been checked-in at f0e292408, but this _breaks the build_ because of this problem.

Reply:
@sydb This all began when @joeytakeda and I were contemplating the valid use of `@generatedBy`. My comment that you did not understand was that your Schematron did not reflect the prose about the `<post>` element. I was also just saying that testing for the ancestor of an attribute seemed problematic for reliable processing, essentially a summary of Joey's point about the skeleton implementation. But the absence of `<post>` in the examples seems problematic too, given what's written in the chapter. 

I think we should add `<post>` to the examples to solve this.